File: python/veval/_context.py
from __future__ import annotations
import inspect
from collections import deque
from typing import Any, Callable, Awaitable, Optional, TypeVar, TYPE_CHECKING
from ._step import Step, StepHandle

if TYPE_CHECKING:
    from ._tracing import TraceData

T = TypeVar("T")


class VevalExecutionContext:
    def __init__(self, trace_id: str, input: Any = None):
        self.trace_id = trace_id
        self.input = input
        self._steps: list[Step] = []
        self._metadata: dict[str, Any] = {}
        self._mock_outputs: Optional[dict[str, deque]] = None
        self._strict_mock_mode = False

    @property
    def steps(self) -> list[Step]:
        return self._steps

    @property
    def trace_meta(self) -> dict[str, Any]:
        return self._metadata

    async def track_step_async(
        self,
        name: str,
        input: Any,
        step: Callable[..., Awaitable[T]],
    ) -> T:
        if self._mock_outputs is not None and self._strict_mock_mode:
            q = self._mock_outputs.get(name)
            if not q:
                available = ", ".join(self._mock_outputs.keys())
                raise RuntimeError(
                    f"Replay mode: no mock output for step '{name}'. "
                    f"Available steps: {available}. "
                    "This would have made a real LLM call."
                )

        if self._mock_outputs is not None:
            q = self._mock_outputs.get(name)
            if q:
                mock = q.popleft()
                s = Step(name)
                s.input = input
                s.metadata["_source"] = "replay"
                s.complete(mock)
                self._steps.append(s)
                return mock  # type: ignore[return-value]

        s = Step(name)
        s.input = input
        self._steps.append(s)
        handle = StepHandle(s)

        try:
            sig = inspect.signature(step)
            if len(sig.parameters) > 0:
                result = await step(handle)
            else:
                result = await step()  # type: ignore[call-arg]
            s.complete(result)
            return result
        except Exception as ex:
            s.fail(str(ex))
            raise

    def set_metadata(self, key: str, value: Any) -> None:
        self._metadata[key] = value

    def load_mock_outputs(self, trace: TraceData, strict: bool = True) -> None:
        if not trace.steps:
            raise RuntimeError(
                f"Replay trace '{trace.trace_id}' has no steps. Cannot mock LLM calls. "
                "Ensure the trace was recorded with steps before using it for replay."
            )
        self._mock_outputs = {}
        self._strict_mock_mode = strict
        for step in trace.steps:
            if step.name not in self._mock_outputs:
                self._mock_outputs[step.name] = deque()
            self._mock_outputs[step.name].append(step.output)

File: node/src/context.ts
import { Step, StepHandle } from "./step";
import { StepData, TraceData } from "./tracing";

export interface JudgeRecord {
  criteria: string;
  score: number;
  passed: boolean;
  reasoning: string;
}

export class VevalExecutionContext {
  readonly traceId: string;
  readonly input: unknown;
  private _steps: Step[] = [];
  private _metadata: Record<string, unknown> = {};
  private _judgments: JudgeRecord[] = [];
  private _mockOutputs: Map<string, StepData[]> | null = null;
  private _strictMockMode = false;

  constructor(traceId: string, input: unknown) {
    this.traceId = traceId;
    this.input = input;
  }

  get steps(): readonly Step[] {
    return this._steps;
  }

  get traceMeta(): Readonly<Record<string, unknown>> {
    return this._metadata;
  }

  get judgments(): readonly JudgeRecord[] {
    return this._judgments;
  }

  setMetadata(key: string, value: unknown): void {
    this._metadata[key] = value;
  }

  recordJudgment(criteria: string, score: number, passed: boolean, reasoning: string): void {
    this._judgments.push({ criteria, score, passed, reasoning });
  }

  async trackStepAsync<T>(
    name: string,
    input: unknown,
    fn: ((handle: StepHandle) => Promise<T>) | (() => Promise<T>)
  ): Promise<T> {
    if (this._mockOutputs !== null && this._strictMockMode) {
      const q = this._mockOutputs.get(name);
      if (!q || q.length === 0) {
        const available = [...(this._mockOutputs?.keys() ?? [])].join(", ");
        throw new Error(
          `Replay mode: no mock output for step '${name}'. ` +
          `Available steps: ${available}. ` +
          "This would have made a real LLM call."
        );
      }
    }

    if (this._mockOutputs !== null) {
      const q = this._mockOutputs.get(name);
      if (q && q.length > 0) {
        const recorded = q.shift()!;
        const mock = recorded.output;
        const s = new Step(name);
        s.input = input;
        // Keep the recorded type so toolCalled and snapshot types behave the same as in the original run.
        s.type = recorded.type ?? "custom";
        s.metadata["_source"] = "replay";
        s.complete(mock);
        this._steps.push(s);
        return mock as T;
      }
    }

    const s = new Step(name);
    s.input = input;
    this._steps.push(s);
    const handle = new StepHandle(s);

    try {
      const result =
        fn.length === 0
          ? await (fn as () => Promise<T>)()
          : await (fn as (h: StepHandle) => Promise<T>)(handle);
      s.complete(result);
      return result;
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      s.fail(msg);
      throw err;
    }
  }

  loadMockOutputs(trace: TraceData, strict = true): void {
    if (!trace.steps || trace.steps.length === 0) {
      throw new Error(
        `Replay trace '${trace.trace_id}' has no steps. Cannot mock LLM calls. ` +
        "Ensure the trace was recorded with steps before using it for replay."
      );
    }
    this._mockOutputs = new Map();
    this._strictMockMode = strict;
    for (const step of trace.steps) {
      if (!this._mockOutputs.has(step.name)) {
        this._mockOutputs.set(step.name, []);
      }
      this._mockOutputs.get(step.name)!.push(step);
    }
  }
}
